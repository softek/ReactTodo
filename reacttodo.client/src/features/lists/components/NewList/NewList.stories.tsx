import type { Meta, StoryObj } from '@storybook/react';
import {
  expect, fn, userEvent, waitFor, within,
} from '@storybook/test';
import { NewList } from './NewList';

const meta = {
  title: 'Features/ListManagement/NewList',
  component: NewList,
  parameters: {
    layout: 'centered',
  },
  argTypes: {
  },
  args: {
    onCreateList: fn(),
  },
} satisfies Meta<typeof NewList>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  play: async ({ args, canvasElement }) => {
    const canvas = within(canvasElement);

    const inputBox = canvas.getByPlaceholderText('New list');
    await expect(inputBox).toBeInTheDocument();

    const createButton = canvas.getByRole('button', { name: /Create List/i });
    await expect(createButton).toBeInTheDocument();

    await userEvent.click(createButton);
    await expect(args.onCreateList).not.toHaveBeenCalled();

    await userEvent.type(inputBox, 'New List Name');
    await userEvent.click(createButton);
    await waitFor(() => expect(args.onCreateList).toHaveBeenCalledWith('New List Name'));
    await expect(inputBox).toHaveValue('');
  },
};
