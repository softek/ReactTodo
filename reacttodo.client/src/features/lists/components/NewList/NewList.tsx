import React from 'react';
import { Button, Input } from 'reactstrap';
import { Controller, useForm } from 'react-hook-form';

type NewListProps = {
  onCreateList: (name: string) => void;
}

type NewListInputs = {
  name: string;
}

export function NewList({ onCreateList: onCreateTask }: NewListProps) {
  const {
    control, handleSubmit, formState: { errors }, reset,
  } = useForm<NewListInputs>({
    defaultValues: { name: '' },
  });

  const onSubmit = (data: { name: string }) => {
    onCreateTask(data.name);
    reset({}, { keepValues: false, keepErrors: false });
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div className="p-0 d-flex mb-4">
        <Controller
          name="name"
          control={control}
          rules={{ required: true }}
          render={({ field }) => (
            <Input
              placeholder="New list"
              invalid={!!errors.name}
              // eslint-disable-next-line react/jsx-props-no-spreading
              {...field}
              autoFocus
              type="text"
            />
          )}
        />
        <Button color="success" onClick={handleSubmit(onSubmit)}>Create List</Button>
      </div>
    </form>
  );
}
