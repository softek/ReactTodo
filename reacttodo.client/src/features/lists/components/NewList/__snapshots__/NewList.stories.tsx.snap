// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Features/ListManagement/NewList Default play-test 1`] = `
<form>
  <div class="p-0 d-flex mb-4">
    <input placeholder="New list"
           name="name"
           type="text"
           class="form-control"
           aria-invalid="false"
           value
    >
    <button type="button"
            class="btn btn-success"
    >
      Create List
    </button>
  </div>
</form>
`;
